src/audio/mixer.c

```diff
@@ -0,0 +1,646 @@
+#include "libdragon.h"
+#include "regsinternal.h"
+#include "mixer.h"
+#include <memory.h>
+#include <stdlib.h>
+#include <math.h>
+#include <stdio.h>
+
+#define MIXER_TRACE   0
+
+#if MIXER_TRACE
+#define tracef(fmt, ...)  debugf(fmt, ##__VA_ARGS__)
+#else
+#define tracef(fmt, ...)  ({ })
+#endif
+
+/**
+ * @name AI Status Register Values
+ * @{
+ */
+/** @brief Bit representing that the AI is busy */
+#define AI_STATUS_BUSY  ( 1 << 30 )
+/** @brief Bit representing that the AI is full */
+#define AI_STATUS_FULL  ( 1 << 31 )
+/** @} */
+
+#define MAX_EVENTS              32
+#define MIXER_POLL_PER_SECOND   8
+
+/**
+ * RSP mixer ucode (rsp_mixer.S)
+ */
+DEFINE_RSP_UCODE(rsp_mixer);
+
+#define MIN(a,b)  ({ typeof(a) _a = a; typeof(b) _b = b; _a < _b ? _a : _b; })
+#define ROUND_UP(n, d)  (((n) + (d) - 1) / (d) * (d))
+
+// NOTE: keep these in sync with rsp_mixer.S
+#define CH_FLAGS_BPS_SHIFT  (3<<0)   // BPS shift value
+#define CH_FLAGS_16BIT      (1<<2)   // Set if the channel is 16 bit
+#define CH_FLAGS_STEREO     (1<<3)   // Set if the channel is stereo (left)
+#define CH_FLAGS_STEREO_SUB (1<<4)   // The channel is the second half of a stereo (right)
+
+// Fixed point value used in waveform position calculations. This is a signed
+// 64-bit integer with the fractional part using MIXER_FX64_FRAC bits.
+// You can use MIXER_FX64() to convert from float.
+typedef uint64_t mixer_fx64_t;
+
+// Fixed point value used for volume and panning calculations.
+// You can use MIXER_FX15() to convert from float.
+typedef int16_t mixer_fx15_t;
+
+
+#define MIXER_FX64_FRAC    12    // NOTE: this must be the same of WAVERFORM_POS_FRAC_BITS in rsp_mixer.S
+#define MIXER_FX64(f)      (int64_t)((f) * (1<<MIXER_FX64_FRAC))
+
+#define MIXER_FX15_FRAC    15
+#define MIXER_FX15(f)      (int16_t)((f) * ((1<<MIXER_FX15_FRAC)-1))
+
+#define MIXER_FX16_FRAC    16
+#define MIXER_FX16(f)      (int16_t)((f) * ((1<<MIXER_FX16_FRAC)-1))
+
+typedef struct mixer_channel_s {
+	/* Current position within the waveform (in bytes) */
+	mixer_fx64_t pos;
+	/* Step between samples (in bytes) to playback at the correct frequency */
+	mixer_fx64_t step;
+	/* Length of the waveform (in bytes) */
+	mixer_fx64_t len;
+	/* Length of the loop in the waveform (in bytes) */
+	mixer_fx64_t loop_len;
+	/* Pointer to the waveform */
+	void *ptr;
+	/* Misc flags */
+	uint32_t flags;
+} mixer_channel_t;
+
+typedef struct rsp_mixer_channel_s {
```

Comment:
There's an assert on the `rsp_mixer_channel_s` struct size. Do we want to use packed to make this explicit?
As they're all 4 byte data datatypes it shouldnt really be an issue, I wouldnt hold anything up with this.

Reply:
Done. I don't think it matters per-se, as long as the RSP definition is the same. That is, even if the non-packed structure had a uint8_t, it would be sufficiente for the RSP DMEM structure to follow it with its packing. OTOH, RSP allows misaligned memory accesses, so packed saves DMEM, doesn't slow down the RSP, and doesn't force the programmer to think of padding rules. So all in all I think you're right that it's better.